certbot-apache/tests/configurator_test.py

```diff
@@ -1752,18 +1762,81 @@ def test_manually_modified_past_file_warns(self):
+        This test ensures that all necessary hashes are present.
+        """
         from certbot_apache._internal.constants import ALL_SSL_OPTIONS_HASHES
-        self.assertTrue(self._current_ssl_options_hash() in ALL_SSL_OPTIONS_HASHES,
-            "Constants.ALL_SSL_OPTIONS_HASHES must be appended"
-            " with the sha256 hash of self.config.mod_ssl_conf when it is updated.")
+        import pkg_resources
+
+        tls_configs_dir = pkg_resources.resource_filename(
+            "certbot_apache", os.path.join("_internal", "tls_configs"))
+        all_files = [os.path.join(tls_configs_dir, name) for name in os.listdir(tls_configs_dir)
+                     if name.endswith('options-ssl-apache.conf')]
+        self.assertTrue(all_files)
+        for one_file in all_files:
+            file_hash = crypto_util.sha256sum(one_file)
+            self.assertTrue(file_hash in ALL_SSL_OPTIONS_HASHES,
+                            "Constants.ALL_SSL_OPTIONS_HASHES must be appended with the sha256 "
+                            "hash of {0} when it is updated.".format(one_file))
+
+    def test_openssl_version(self):
+        some_string_contents = """
+            SSLOpenSSLConfCmd
+            OpenSSL configuration command
+            SSLv3 not supported by this version of OpenSSL
+            '%s': invalid OpenSSL configuration command
+            OpenSSL 1.0.2g  1 Mar 2016
+            OpenSSL
+            AH02407: "SSLOpenSSLConfCmd %s %s" failed for %s
+            AH02556: "SSLOpenSSLConfCmd %s %s" applied to %s
+            OpenSSL 1.0.2g  1 Mar 2016
+            """
+        self.config.parser.modules['ssl_module'] = '/fake/path'
+        with mock.patch("%s.open" % six.moves.builtins.__name__,
+            mock.mock_open(read_data=some_string_contents)) as mock_file:
+            self.assertEqual(self.config.openssl_version(), "1.0.2g")
+
+    def test_current_version(self):
+        self.config.version = (2, 4, 10)
+        self.config._openssl_version = '1.0.2m'
+        self.assertTrue('old' in self.config.pick_apache_config())
+
+        self.config.version = (2, 4, 11)
+        self.config._openssl_version = '1.0.2m'
+        self.assertTrue('current' in self.config.pick_apache_config())
+
+        self.config._openssl_version = '1.0.2a'
+        self.assertTrue('old' in self.config.pick_apache_config())
+
+    def test_openssl_version_errors(self):
+        self.config._openssl_version = '1.0.2a'
+        self.assertEqual(self.config.openssl_version(), '1.0.2a')
+
+        self.config._openssl_version = None
+        self.config.parser.modules['ssl_module'] = None
+        self.assertEqual(self.config.openssl_version(), None)
+
+        self.config.parser.modules['ssl_module'] = "/fake/path"
+        self.assertRaises(errors.PluginError, self.config.openssl_version)
+
+        contents_missing_openssl = "these contents won't match the regex"
+        with mock.patch("%s.open" % six.moves.builtins.__name__,
```

Comment:
I found this approach somewhere else that I sure do wish I'd kept the reference to now; I'll switch to the mock version.

Reply:
I guess there could theoretically be a different between `configurator` and `configurator_test`'s mocks, but probably  not, and if it there, it's better to use `configurator`'s.

Reply:
we also apparently use the builtins approach in other tests above.

Reply:
well, this is moot now anyway.

Reply:
it's no longer moot! but our `oldest` tests don't like this, so I'm changing it back because apparently there is some reason that the other way works.

Reply:
Sorry for the bad suggestion!

I couldn't stop myself from digging into this so if you're also curious, what's going on here is newer versions of `mock` will automatically let you patch out builtin objects, while older ones require you pass `create=True` to `mock.patch` because they fail to find the builtin object to be patched since the lookups for these objects are done differently. There's a note about this buried in the [mock.patch docs](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.patch).

Reply:
yeah, this seems to be an idiomatic way to do it anyway.

Reply:
thanks for investigating! I was mildly interested in knowing why but not enough to look it up myself.